server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -0,0 +1,240 @@
+      } catch {
+        // We cannot kill the YARN app without the app id.
+        // There's a chance the YARN app hasn't been submitted during a livy-server failure.
+        // We don't want a stuck session that can't be deleted. Emit a warning and move on.
+        case _: TimeoutException | _: InterruptedException =>
+          warn("Deleting a session while its YARN application is not found.")
+          yarnAppMonitorThread.interrupt()
+      }
+    }
+  }
+
+  private def changeState(newState: SparkApp.State.Value): Unit = {
+    if (state != newState) {
+      listener foreach (_.stateChanged(state, newState))
+      state = newState
+    }
+  }
+
+  private def getYarnDiagnostics(appReport: ApplicationReport): IndexedSeq[String] = {
+    Option(appReport.getDiagnostics)
+      .filter(_.nonEmpty)
+      .map[IndexedSeq[String]]("YARN Diagnostics:" +: _.split("\n"))
+      .getOrElse(IndexedSeq.empty)
+  }
+
+  private def isRunning: Boolean = {
+    state != SparkApp.State.FAILED && state != SparkApp.State.FINISHED &&
+    state != SparkApp.State.KILLED
+  }
+
+  // Exposed for unit test.
+  private[utils] def mapYarnState(
+      appId: ApplicationId,
+      yarnAppState: YarnApplicationState,
+      finalAppStatus: FinalApplicationStatus): SparkApp.State.Value = {
+    yarnAppState match {
+      case (YarnApplicationState.NEW |
+            YarnApplicationState.NEW_SAVING |
+            YarnApplicationState.SUBMITTED |
+            YarnApplicationState.ACCEPTED) => SparkApp.State.STARTING
+      case YarnApplicationState.RUNNING => SparkApp.State.RUNNING
+      case YarnApplicationState.FINISHED =>
+        finalAppStatus match {
+          case FinalApplicationStatus.SUCCEEDED => SparkApp.State.FINISHED
+          case FinalApplicationStatus.FAILED => SparkApp.State.FAILED
+          case FinalApplicationStatus.KILLED => SparkApp.State.KILLED
+          case s =>
+            error(s"Unknown YARN final status $appId $s")
+            SparkApp.State.FAILED
+        }
+      case YarnApplicationState.FAILED => SparkApp.State.FAILED
+      case YarnApplicationState.KILLED => SparkApp.State.KILLED
+    }
+  }
+
+  // Exposed for unit test.
+  // TODO Instead of spawning a thread for every session, create a centralized thread and
+  // batch YARN queries.
+  private[utils] val yarnAppMonitorThread = new Thread(s"yarnAppMonitorThread-$this") {
+    override def run() = {
+      @tailrec
+      def waitForAppId(): ApplicationId = {
+        try {
+          // Wait for spark-submit to finish submitting the app to YARN.
+          process map (_.waitFor()) match {
+            case (None | Some(0)) => Await.result(appIdFuture, SparkYarnApp.POLL_INTERVAL)
+            case Some(exitCode) =>
+              throw new Exception(s"spark-submit exited with code $exitCode}.\n" +
+                s"${process.get.inputLines mkString "\n"}")
+          }
+        } catch {
+          case e: TimeoutException => waitForAppId()
```

Comment:
Not sure I understand why you have a timeout if you're calling the method recursively once the `Await` call times out?